File: 0201-0300/0295.find-median-from-data-stream.cpp
// TODO: 其他题解方法
/*
 * 295. 数据流的中位数 - Hard
 * 中位数是有序列表中间的数。如果列表长度是偶数，中位数则是中间两个数的平均值。

例如，

[2,3,4] 的中位数是 3

[2,3] 的中位数是 (2 + 3) / 2 = 2.5

设计一个支持以下两种操作的数据结构：

void addNum(int num) - 从数据流中添加一个整数到数据结构中。
double findMedian() - 返回目前所有元素的中位数。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/find-median-from-data-stream
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */


class MedianFinder {
public:
    /** initialize your data structure here. */
    MedianFinder() {

    }
    
    void addNum(int num) {
        if (smaller_q.empty()) {
            smaller_q.push(num);
            return;
        }
        // larger_q.size() <= smaller_q.size() <= larger_q.size() + 1
        if (num <= smaller_q.top()) {
            smaller_q.push(num);
            if (smaller_q.size() > larger_q.size() + 1) {
                larger_q.push(smaller_q.top());
                smaller_q.pop();
            }
        } else {
            larger_q.push(num);
            if (larger_q.size() > smaller_q.size()) {
                smaller_q.push(larger_q.top());
                larger_q.pop();
            }
        }
    }
    
    double findMedian() {
        if (smaller_q.size() != larger_q.size())
            return smaller_q.top();
        return (smaller_q.top() + larger_q.top()) / 2.0;
    }
private:
    priority_queue<int> smaller_q;
    priority_queue<int, vector<int>, greater<int> > larger_q;
    // smaller_q, larger_q
};

File: 0201-0300/0295-find-median-from-data-stream.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

#[derive(Default)]
struct MedianFinder {
    left: BinaryHeap<i32>,  // maximum heap
    right: BinaryHeap<Reverse<i32>>,  // minimum heap
}

impl MedianFinder {

    /** initialize your data structure here. */
    fn new() -> Self {
        Default::default()
    }
    
    fn add_num(&mut self, num: i32) {
        self.left.push(num);
        self.right.push(Reverse(self.left.pop().unwrap()));
        if self.right.len() > self.left.len() {
            self.left.push(self.right.pop().unwrap().0);
        }
    }
    
    fn find_median(&self) -> f64 {
        if self.left.len() == self.right.len() {
            return (*self.left.peek().unwrap() as f64 + self.right.peek().unwrap().0 as f64) / 2.0; 
        }
        *self.left.peek().unwrap() as f64
    }
}

fn main() {
    let mut mf = MedianFinder::new();
    mf.add_num(1);
    mf.add_num(2);
    assert_eq!(mf.find_median(), 1.5);
    mf.add_num(3);
    assert_eq!(mf.find_median(), 2.0);
}
